runtime/src/serde_snapshot/newer.rs

```diff
@@ -460,18 +519,102 @@ enum BankFieldsToSerialize<'a, T: Serialize> {
+    // This allows this code to be present and harmless
+    WithVersionedEpochStakes(
+        (
+            SerializableVersionedBank<'a>,
+            T,
+            u64,
+            Option<BankIncrementalSnapshotPersistence>,
+            Option<Hash>,
+            UnusedEnum,
+            HashMap<Epoch, VersionedEpochStakes>,
+        ),
+    ),
+}
+
+fn split_epoch_stakes(
+    bank_epoch_stakes: &mut HashMap<Epoch, EntryStakes>,
+) -> (
+    HashMap<Epoch, EntryStakes>,
+    HashMap<Epoch, VersionedEpochStakes>,
+) {
+    let mut old_epoch_stakes = HashMap::new();
+    let mut new_epoch_stakes = HashMap::new();
+    for (epoch, epoch_stakes) in bank_epoch_stakes.drain() {
+        let EpochStakes {
+            stakes,
+            total_stake,
+            node_id_to_vote_accounts,
+            epoch_authorized_voters,
+        } = epoch_stakes;
+
+        match stakes.as_ref() {
+            StakesEnum::Delegations(_) => {
+                old_epoch_stakes.insert(
+                    epoch,
+                    EpochStakes {
+                        stakes: stakes.clone(),
+                        total_stake,
+                        node_id_to_vote_accounts,
+                        epoch_authorized_voters,
+                    },
+                );
+            }
+            StakesEnum::Accounts(stakes) => {
+                new_epoch_stakes.insert(
+                    epoch,
+                    VersionedEpochStakes::Current {
+                        stakes: Stakes::<Stake>::from(stakes.clone()),
+                        total_stake,
+                        node_id_to_vote_accounts,
+                        epoch_authorized_voters,
+                    },
+                );
+            }
+            StakesEnum::Stakes(stakes) => {
+                new_epoch_stakes.insert(
+                    epoch,
+                    VersionedEpochStakes::Current {
+                        stakes: stakes.clone(),
+                        total_stake,
+                        node_id_to_vote_accounts,
+                        epoch_authorized_voters,
+                    },
+                );
+            }
+        }
+    }
+    (old_epoch_stakes, new_epoch_stakes)
 }
 
 /// serializing involves building these fields into a tuple
 /// This occurs during normal serialization and again during re-serialization.
 fn get_serialize_bank_fields<'a, T: Serialize>(
-    bank: SerializableVersionedBank<'a>,
+    mut bank: SerializableVersionedBank<'a>,
     accounts_db_fields: T,
     lamports_per_signature: u64,
     incremental_snapshot_persistence: Option<BankIncrementalSnapshotPersistence>,
     epoch_accounts_hash: Option<Hash>,
     epoch_reward_status: Option<&'a EpochRewardStatus>,
+    use_new_epoch_stakes_field: bool,
 ) -> BankFieldsToSerialize<'a, T> {
+    if use_new_epoch_stakes_field {
+        let (old_epoch_stakes, new_epoch_stakes) = split_epoch_stakes(&mut bank.epoch_stakes);
```

Comment:
Before this change, bank `epoch_stakes` could have entries with `StakesEnum::Accounts` (inserted at epoch boundaries) and `StakesEnum::Delegations` (deserialized from snapshots). When the feature in the change is activated, snapshots will start containing `StakesEnum::Stakes` (via `VersionedEpochStakes`) but there could still be some `StakesEnum::Delegations` entries for awhile for nodes that restarted recently. So in order for partitioned rewards to be sure that an entry in `bank.epoch_stakes` for a particular epoch uses `StakesEnum::Stakes` we will need to wait a few epochs.